### PR TITLE
feat: Add settings card to control link previews settings in chat input

### PR DESCRIPTION
**Pull request by alexjba**
### What does the PR do
Closing: #12180

Depends on #12262 to be merged first.

This commit adds the link preview settings card in the chat input area and connects the settings to the controller.

Not included in this commit: Backend for the preserving the settings, syncing the settings and enforcing the settings on the backend side.

Whenever an url is detected in the chat input area, the link preview settings card is presented. This card enables the user to choose one of the following options:

1. `Show for this message` - All the link previews in the current message will be loaded without asking again. The current message can be defined as the message currently typed/pasted in the chat input. Deleting or sending the current content is resetting this setting and the link preview settings card will be presented again when a new url is detected.
2. `Always show previews` - All the link previews will be loaded automatically. The link preview settings card will not be presented again (in the current state, this settings is enabled for the lifetime of the controller. This will change once the settings are preserved and synced)
3. `Never show previews` - No link preview will be loaded. Same as the `Always show previews` option, this will be preserved for the lifetime of the controller for now.
4. Dismiss (x button) - The link preview settings card will be dismissed. It will be loaded again when a new link preview is detected

The same options can be loaded as a context menu on the link preview card.

Changes:
1. Adding `LinkPreviewSettingsCard`
2. Adding the settings context menu to `LinkPreviewSettingsCard` and `LinkPreviewMiniCard`
3. Connect settings events to the nim controller
4. Adding the controller logic for settings change
5. Adding the link preview dismiss settings flag to the preserverd properties and use it as a condition to load the settings.
6. Adding/Updating corresponding storybook pages

### Affected areas

Link previews
StatusChatInput
Storybook

#### New url detected

https://github.com/status-im/status-desktop/assets/47811206/a901d56b-b2c1-46c4-9969-571de00d0c33

#### Dismiss settings

https://github.com/status-im/status-desktop/assets/47811206/6270be5b-6c1f-4912-a258-e91d4cd90e8f

#### Enable for current message

https://github.com/status-im/status-desktop/assets/47811206/f5cdfa73-af18-4c67-a74c-834bd227734b

#### Enable (partial implementation - needs backend for persistent settings. Currently this setting is preserved for the controller lifetime)

https://github.com/status-im/status-desktop/assets/47811206/c5606180-97da-4031-a243-c6b62bd1cc59

#### Disable

https://github.com/status-im/status-desktop/assets/47811206/b069fdc2-1af1-4361-91a9-b2c3498bf3c9

#### Disable from the link preview card

https://github.com/status-im/status-desktop/assets/47811206/7c29d3b6-cf92-45b2-a9b7-2c97b0bc5098

- [x] I've checked the design and this PR matches it

### Cool Spaceship Picture

<!-- optional but cool ->

### CODE REVIEW

**Review by igor-sirotin — COMMENTED** [review]
The code looks good to me, thank you!
I didn't check the logic very careful. Will be much easier to do that when connecting to the settings backend 👌 
________

I have a small concern of the UI though.

**Comment by igor-sirotin** [comment]
Also I can see some differences in the UI of the settings card.
- content is slightly vertically misaligned
- padding between title and subtitle is too big
- padding between buttons is too big
- I think card paddings are bigger in UI (? not sure)
- menu should appear above the button, not overlapping it

### Figma:
image

### Implementation:
Screenshot 2023-10-09 at 22 01 32

File: ui/imports/shared/controls/chat/LinkPreviewSettingsCard.qml [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    borderWidth: 0
+    verticalPadding: 12
+    horizontalPadding: Style.current.padding
+
+    onContextMenuChanged: {
+        if (!contextMenu) {
+            console.error("No context menu provided")
+            return
+        }
+        contextMenu.parent = optionsComboBox
```
[the remark this answers is not in the document]
Comment (alexjba) [reply ui/imports/shared/controls/chat/LinkPreviewSettingsCard.qml:30]: > Especially that you already created `enableLinkPreviews` and `disableLinkPreviews` signals that are unused here 😄

Ohh, that's some leftovers. Also the reparenting thing. I'll remove it. Thanks for spotting that! 👍 
I used to have a single context menu item shared between settings card and link previews. But didn't work very well.

File: src/app/modules/main/chat_section/chat_content/input_area/controller.nim [hunk 2]
```diff
@@ -154,11 +176,25 @@ proc addToRecentsGif*(self: Controller, item: GifDto) =
...
   let urls = self.messageService.getTextUrls(text)
   self.delegate.setUrls(urls)
-  if len(urls) > 0:
-    let newUrls = self.linkPreviewCache.unknownUrls(urls)
+  let newUrls = self.linkPreviewCache.unknownUrls(urls)
+
+  let askToEnableLinkPreview = len(newUrls) > 0 and self.canAskToEnableLinkPreview()
+  self.delegate.setAskToEnableLinkPreview(askToEnableLinkPreview)
+
+  if self.getLinkPreviewEnabled() == true and len(urls) > 0:
```
[the remark this answers is not in the document]
Comment (igor-sirotin) [reply src/app/modules/main/chat_section/chat_content/input_area/controller.nim:197]: There several such places in this PR, if you don't mind 🙂

File: ui/imports/shared/controls/chat/LinkPreviewSettingsCard.qml [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+                StatusBaseText {
+                    Layout.fillWidth: true
+                    Layout.fillHeight: true
+                    verticalAlignment: Text.AlignVCenter
+                    elide: Text.ElideRight
+                    text: qsTr("Options")
+                    color: Theme.palette.baseColor1
+                }
+                StatusIcon {
+                    width: 18
```
Comment (noeliaSD) [inline ui/imports/shared/controls/chat/LinkPreviewSettingsCard.qml:78]:
```suggestion
                    Layout.preferredWidth: 18
```

File: ui/imports/shared/controls/chat/ChatInputLinksPreviewArea.qml [hunk 4]
```diff
@@ -4,26 +4,35 @@ import QtGraphicalEffects 1.15
...
 Control {
     id: root
 
     required property var imagePreviewModel
     required property var linkPreviewModel
+    required property bool showLinkPreviewSettings
 
     readonly property alias hoveredUrl: d.hoveredUrl
-    readonly property int contentItemsCount: imagePreviewModel.length + d.filteredModel.count
+    readonly property bool hasContent: imagePreviewModel.length || showLinkPreviewSettings || linkPreviewRepeater.count
```
Comment (noeliaSD) [inline ui/imports/shared/controls/chat/ChatInputLinksPreviewArea.qml:24]: More explicit?

```suggestion
    readonly property bool hasContent: imagePreviewModel.length > 0 || showLinkPreviewSettings || linkPreviewRepeater.count
```

**Review by noeliaSD — COMMENTED** [review]
In general LGTM! Just added some comments and found an issue while testing:

- The existing implementation does not behave as written in this figma comment: `When “X” is clicked, no link preview prompt should be shown for subsequent links entered into the current message but it should be shown as soon as a link is entered into a subsequent message`.
Now, every time a new link is pasted into the current message, the preview card is shown again although `X` was previously clicked.

File: ui/imports/shared/controls/chat/ChatInputLinksPreviewArea.qml (hunk 4 above)
Comment (caybro) [reply ui/imports/shared/controls/chat/ChatInputLinksPreviewArea.qml:24]: It's JavaScript, `linkPreviewRepeater.count` is fine? ;)

File: src/app/modules/main/chat_section/chat_content/input_area/controller.nim [hunk 5]
```diff
@@ -22,6 +29,8 @@ type
     gifService: gif_service.Service
     messageService: message_service.Service
     linkPreviewCache: LinkPreviewCache
+    linkPreviewPersistentSetting: LinkPreviewSetting
+    linkPreviewTemporarySetting: LinkPreviewSetting
```
[the remark this answers is not in the document]
Comment (alexjba) [reply src/app/modules/main/chat_section/chat_content/input_area/controller.nim:33]: > but it really keeps the state of the current message,

Yes, you're right!
I was thinking it's suggestive enough since we have `linkPreviewPersistentSetting` and `linkPreviewTemporarySetting`

`linkPreviewCurrentMessageSetting` could be better actually since this controller now has some logic that's determining the lifetime of this settings as being equal to the current message.

**Comment by alexjba** [comment]
@igor-sirotin @noeliaSD Please have a second look. I've updated the PR

**Review by igor-sirotin — APPROVED** [review]
Thanks, @alexjba 🎉 
The code LGTM, but I didn't test it 🙂

**Review by noeliaSD — APPROVED** [review]
LGTM! Also tested!

File: storybook/pages/LinkPreviewSettingsCardPage.qml [hunk 6]
```diff
@@ -0,0 +1,43 @@
+import QtQuick 2.15
+import QtQuick.Controls 2.15
+import QtQuick.Layouts 1.15
+
+import Qt.labs.settings 1.0
```
Comment (micieslak) [inline storybook/pages/LinkPreviewSettingsCardPage.qml:5]: please remove unused imports

File: ui/imports/shared/controls/chat/LinkPreviewSettingsCard.qml [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+            Layout.preferredHeight: 38
+            Layout.preferredWidth: 38
+            type: StatusFlatRoundButton.Type.Secondary
+            icon.name: "close"
+            icon.color: Theme.palette.directColor1
+            onClicked: root.dismiss()
+        }
+    }
+
+    component ContextMenu: StatusMenu {
```
Comment (micieslak) [inline ui/imports/shared/controls/chat/LinkPreviewSettingsCard.qml:110]: When there is only one usage of that component, is it easier to keep it inline? The signals could be emitted directly resulting with simpler and shorter code.

File: ui/imports/shared/controls/chat/ChatInputLinksPreviewArea.qml [hunk 8]
```diff
@@ -107,6 +117,7 @@ Control {
                         onClose: root.linkPreviewModel.removePreviewData(d.filteredModel.mapToSource(index))
```
Comment (micieslak) [inline ui/imports/shared/controls/chat/ChatInputLinksPreviewArea.qml:117]: I know it's not part of that pr, but assuming some methods on models is quite unusual and bit problematic. Cleaner solution is to emit signal here (like e.g. `previewDataRemoveRequested`, called with mapped index) instead of calling `removePreviewData` and handling removal externally.

Even now, in storybook we have:
```
file:///home/michalc/workspace/status-desktop/ui/imports/shared/controls/chat/ChatInputLinksPreviewArea.qml:117: TypeError: Property 'removePreviewData' of object QQmlListModel(0x55cd0fb11270) is not a function
```
when clicking "X" on an item

File: ui/imports/shared/controls/chat/ChatInputLinksPreviewArea.qml [hunk 9]
```diff
@@ -8,22 +8,29 @@ import StatusQ.Core 0.1
...
 import shared.controls.chat 1.0
 
+import utils 1.0
+
 import SortFilterProxyModel 0.2
 
 Control {
     id: root
 
     required property var imagePreviewModel
```
Comment (micieslak) [inline ui/imports/shared/controls/chat/ChatInputLinksPreviewArea.qml:18]: Not part of this pr, but what is expected here is not a model but js array, what is quite misleading from component's api perspective. It would be nice to rename it or add appropriate comment.

File: ui/imports/shared/controls/chat/ChatInputLinksPreviewArea.qml [hunk 10]
```diff
@@ -8,22 +8,29 @@ import StatusQ.Core 0.1
...
     required property var linkPreviewModel
```
Comment (micieslak) [inline ui/imports/shared/controls/chat/ChatInputLinksPreviewArea.qml:19]: It would be cool to have expected roles listed in a comment.

**Review by micieslak — APPROVED** [review]
LGTM in general, some minor comments only. Nice job!

File: ui/imports/shared/controls/chat/LinkPreviewSettingsCard.qml (hunk 7 above)
Comment (alexjba) [reply ui/imports/shared/controls/chat/LinkPreviewSettingsCard.qml:110]: It's reused for the other cards and this is why I'm exposing it. The same settings menu will be opened when right clicking a link preview.
